src/main/java/com/mraof/minestuck/tileentity/redstone/StructureCoreTileEntity.java

```diff
@@ -0,0 +1,256 @@
+	 * READ AND REDSTONE checks to see if the structure piece hasBeenCompleted variable is true, and then gives off a redstone signal if so
+	 * WRITE turns the structure piece hasBeenCompleted variable to true if it is not yet
+	 */
+	public enum ActionType
+	{
+		READ_AND_WIPE,
+		READ_AND_REDSTONE,
+		WRITE;
+		
+		public static ActionType fromInt(int ordinal) //converts int back into enum
+		{
+			for(ActionType type : ActionType.values())
+			{
+				if(type.ordinal() == ordinal)
+					return type;
+			}
+			throw new IllegalArgumentException("Invalid ordinal of " + ordinal + " for structure core action type!");
+		}
+		
+		public String getNameNoSpaces()
+		{
+			return name().replace('_', ' ');
+		}
+	}
+	
+	public StructureCoreTileEntity()
+	{
+		super(MSTileEntityTypes.STRUCTURE_CORE.get());
+		actionType = ActionType.READ_AND_REDSTONE;
+	}
+	
+	@Override
+	public void tick()
+	{
+		if(level == null || level.isClientSide)
+			return;
+		
+		int cycleRate = (hasWiped && actionType == ActionType.READ_AND_WIPE) ? 600 : 100;
+		if(tickCycle >= cycleRate)
+		{
+			sendUpdate();
+			tickCycle = 0;
+		}
+		tickCycle++;
+	}
+	
+	private void sendUpdate()
+	{
+		if(level != null && level.isAreaLoaded(getBlockPos(), 1) && getBlockState().getValue(StructureCoreBlock.ACTIVE))
+		{
+			StructureManager structureManager = ((ServerWorld) level).structureFeatureManager();
+			StructureStart<?> structureStart = getStructureStart(structureManager);
+			if(structureStart != null && structureStart.isValid())
+			{
+				Debug.debugf("meets all conditions, now doing action type function");
+				if(actionType == ActionType.WRITE)
+				{
+					writeToStructure(structureStart);
+				} else if(actionType == ActionType.READ_AND_WIPE)
+				{
+					wipeSlate();
+				} else if(actionType == ActionType.READ_AND_REDSTONE)
+				{
+					CoreCompatibleScatteredStructurePiece piece = getStructurePiece(structureStart);
+					
+					if(piece != null && piece.hasBeenCompleted())
+					{
+						if(!getBlockState().getValue(StructureCoreBlock.POWERED))
+							level.setBlockAndUpdate(getBlockPos(), getBlockState().setValue(StructureCoreBlock.POWERED, true));
+					} else
+					{
+						level.setBlockAndUpdate(getBlockPos(), getBlockState().setValue(StructureCoreBlock.POWERED, false));
+					}
+				}
+			}
+		}
+	}
+	
+	public StructureStart<?> getStructureStart(StructureManager structureManager)
+	{
+		for(Structure<?> structureListIterate : SOLVABLE_STRUCTURES)
+		{
+			return structureManager.getStructureAt(getBlockPos(), true, structureListIterate);
+		}
+		
+		return null;
+	}
+	
+	private void writeToStructure(StructureStart<?> structureStart)
+	{
+		//compoundnbt.putString("id", Registry.STRUCTURE_FEATURE.getKey(this.getFeature()).toString());
+		
+		CoreCompatibleScatteredStructurePiece piece = getStructurePiece(structureStart);
+		if(piece != null)
+		{
+			if(!piece.hasBeenCompleted())
+				piece.nowCompleted();
+			else
+				Debug.debugf("has already been completed");
```

Comment:
Usages of `Debug` are being replaced. Remove it if you're done with it, or replace it by a local logger.

Reply:
Hope I used the right one

Reply:
Actually ended up removing them, clogged things up